eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/NodeRelayerSpec.scala

```diff
@@ -0,0 +1,561 @@
+import fr.acinq.eclair.payment.PaymentRequest.{ExtraHop, PaymentRequestFeatures}
+import fr.acinq.eclair.payment.send.MultiPartPaymentLifecycle.{PreimageReceived, SendMultiPartPayment}
+import fr.acinq.eclair.payment.send.PaymentInitiator.SendPaymentConfig
+import fr.acinq.eclair.payment.send.PaymentLifecycle.SendPayment
+import fr.acinq.eclair.payment._
+import fr.acinq.eclair.router.Router.RouteRequest
+import fr.acinq.eclair.router.{BalanceTooLow, RouteNotFound}
+import fr.acinq.eclair.wire._
+import fr.acinq.eclair.{CltvExpiry, CltvExpiryDelta, LongToBtcAmount, MilliSatoshi, NodeParams, ShortChannelId, TestConstants, nodeFee, randomBytes, randomBytes32, randomKey}
+import org.scalatest.funsuite.FixtureAnyFunSuiteLike
+import org.scalatest.{Outcome, Tag}
+import scodec.bits.HexStringSyntax
+
+import scala.concurrent.duration._
+import scala.util.Random
+
+/**
+ * Created by t-bast on 10/10/2019.
+ */
+
+class NodeRelayerSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("application")) with FixtureAnyFunSuiteLike {
+
+  import NodeRelayerSpec._
+
+  case class FixtureParam(nodeParams: NodeParams, nodeRelayer: ActorRef[NodeRelay.Command], router: TestProbe[Any], register: TestProbe[Any], mockPayFSM: TestProbe[Any], eventListener: TestProbe[PaymentEvent])
+
+  override def withFixture(test: OneArgTest): Outcome = {
+    eventually {
+      val nodeParams = TestConstants.Bob.nodeParams.copy(multiPartPaymentExpiry = 5 seconds)
+      val router = TestProbe[Any]("router")
+      val register = TestProbe[Any]("register")
+      val eventListener = TestProbe[PaymentEvent]("event-listener")
+      system.eventStream ! EventStream.Subscribe(eventListener.ref)
+      val mockPayFSM = TestProbe[Any]("pay-fsm")
+      val fsmFactory = if (test.tags.contains("mock-fsm")) {
+        new NodeRelay.FsmFactory {
+          override def spawnOutgoingPayFSM(context: ActorContext[NodeRelay.Command], nodeParams: NodeParams, router: akka.actor.ActorRef, register: akka.actor.ActorRef, cfg: SendPaymentConfig, multiPart: Boolean): akka.actor.ActorRef = {
+            mockPayFSM.ref ! cfg
+            mockPayFSM.ref.toClassic
+          }
+        }
+      } else {
+        new NodeRelay.FsmFactory {
+          override def spawnOutgoingPayFSM(context: ActorContext[NodeRelay.Command], nodeParams: NodeParams, router: akka.actor.ActorRef, register: akka.actor.ActorRef, cfg: SendPaymentConfig, multiPart: Boolean): akka.actor.ActorRef = {
+            val fsm = super.spawnOutgoingPayFSM(context, nodeParams, router, register, cfg, multiPart)
+            mockPayFSM.ref ! fsm
+            fsm
+          }
+        }
+      }
+      val nodeRelay = testKit.spawn(NodeRelay(nodeParams, router.ref.toClassic, register.ref.toClassic, relayId, paymentHash, fsmFactory))
+      withFixture(test.toNoArgTest(FixtureParam(nodeParams, nodeRelay, router, register, mockPayFSM, eventListener)))
+    }
+  }
+
+  test("fail to relay when incoming multi-part payment times out") { f =>
```

Comment:
I think this should use a `mockFsm`, otherwise we're waiting for a timeout and this test takes 5s whereas it could take 100ms :)

Reply:
I disagree: this allows us to test the actual timeout logic, and tests are parallelized anyway.

Reply:
Ok, it's already tested in the multi-part FSM spec but it never hurts to test multiple times!